Fix encoding entities

KarolDawidziuk opened the pull request:
<!--
🚨 If you want to submit a PR for a security vulnerability, please contact us directly
at https://ckeditor.com/contact/ instead. 🚨
-->
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#4941](https://github.com/ckeditor/ckeditor4/issues/4941): Fix: Some entities get wrongly encoded, when using `entities_processNumerical = true`
```

## What changes did you make?

There was an issue with using `String#charCodeAt()` in the `entities` plugin that returns an integer from 0 to 65535 representing UTF-16 code unit at the given index. In the case when index was greater than 65535 method returns a surrogate pair. To fix this I used `codePointAt()` which returns a Unicode code point value at the given index.

Another issue was with the regex which was unable to match and replace given unicodes. To fix this I updated existing entities regex by adding a `u`- unicode option. 

Unfortunately, this is not working on IE because `codePointAt()` and Unicode option are not supported there. Although it is possible to correctly convert HTMLEntity based on the values from the surrogate pair(see [d148812](https://github.com/ckeditor/ckeditor4/commit/d148812ee33f5231e4193a1a2acbd96f3f81e489#diff-d14edc554ac7a26f0fe8a2787c381907902340da10665f40992a0e417dfb240aR113-R137)) I was unable to find a solution to support the Unicode option on IE.


## Which issues does your PR resolve?

Closes #4941.

sculpt0r reviewed (changes requested):
@KarolDawidziuk - I need to ask you to add some tests. I didn't look at the code much more (but we discussed it ;) ).

We missed tests (since the beginning, not only for this PR) for [`force`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_config.html#cfg-entities_processNumerical) option 🤔 It is mentioned in docs but our tests only use `true` value. So before we touch the code even more - let's establish the current logic with tests 🤞

KarolDawidziuk commented:
@sculpt0r I've added more unit tests with `force` option and also I added a separate manual test to cover that case and PR is ready for another check.

KarolDawidziuk commented:
@sculpt0r I've added more unit tests with `force` option and also I added a separate manual test to cover that case and PR is ready for another check.

sculpt0r commented:
Rebased on the newest master.

sculpt0r commented:
> I've added more unit tests with force option
Looks good - thank you for that :)

sculpt0r commented on plugins/entities/plugin.js:
```diff
@@ -143,7 +148,9 @@
-				entitiesRegex = new RegExp( entitiesRegex, 'g' );
+
+				// IE does not support unicode option in the Regex constructor (#4941).
+				entitiesRegex = CKEDITOR.env.ie ? new RegExp( entitiesRegex, 'g' ) : new RegExp( entitiesRegex, 'gu' );
```
Lets make this code less repeatable:
```suggestion
				entitiesRegex = new RegExp( entitiesRegex, CKEDITOR.env.ie ? 'g' : 'gu' );
```

sculpt0r commented on tests/plugins/entities/entities.js:
```diff
@@ -82,5 +82,115 @@ bender.test( {
+		} );
+	},
+
+	// (#4941)
```
Let's remove issue reference from tests that were added just for sake of this feature coverage. So this one and  the two others with `force` param.

sculpt0r commented on tests/plugins/entities/manual/entitiesprocessnumericalforce.md:
```diff
@@ -0,0 +1,10 @@
+@bender-tags: bug, 4.20.0, 4941
```
I wouldn't use issue tag here. This test is just for sake of coverage entities feature.

KarolDawidziuk replied:
I think that it's correct now because even if it's for coverage purposes, it's still broken while using `force` option, and this PR fixing it. WDYT?

sculpt0r reviewed (changes requested):
@KarolDawidziuk - FF, Safari, Chrome looks good 👍 IE also is not blown up, so another 👍 

Please take a look at my final inline comments. Sorry for dragging you through writing tests not explicitly related to the issue itself, but we never know when we come back here. So if we have already deep dive into this plugin - we might add tests that will cover this feature for future changes - hope you understand 😓

KarolDawidziuk commented:
Thanks, @sculpt0r for the review.

I've applied changes mentioned in your comments except a change with bender tags in the manual test and PR is ready for another check.

sculpt0r reviewed (approved):
Thanks for all new test cases 👍 

LGTM 🎉
